File: 0460. LFU Cache/Solution.cpp
#include <list>
#include <unordered_map>
using namespace std;

struct Node {
    int key, val, freq;
    Node(int _key, int _val, int _freq)
        : key(_key)
        , val(_val)
        , freq(_freq)
    {
    }
};

class LFUCache {
public:
    LFUCache(int _capacity)
        : capacity(_capacity)
    {
    }

    int get(int key)
    {
        auto it = key2node.find(key);
        if (it == key2node.end())
            return -1;
        list<Node>::iterator node = it->second;
        int val = node->val, freq = node->freq;
        freq2list_del_node(node);
        freq2list[freq + 1].emplace_front(key, val, freq + 1);
        key2node[key] = freq2list[freq + 1].begin();
        return val;
    }

    void put(int key, int value)
    {
        auto it = key2node.find(key);
        if (it == key2node.end()) {
            if (key2node.size() == capacity) {
                Node node = freq2list[min_freq].back();
                freq2list[min_freq].pop_back();
                key2node.erase(node.key);
                if (freq2list[min_freq].empty())
                    freq2list.erase(min_freq);
            }
            freq2list[1].emplace_front(key, value, 1);
            key2node[key] = freq2list[1].begin();
            min_freq = 1;
        } else {
            list<Node>::iterator node = it->second;
            int freq = node->freq;
            freq2list_del_node(node);
            freq2list[freq + 1].emplace_front(key, value, freq + 1);
            key2node[key] = freq2list[freq + 1].begin();
        }
    }

private:
    void freq2list_del_node(list<Node>::iterator it)
    {
        int freq = it->freq;
        freq2list[freq].erase(it);
        if (freq2list[freq].empty()) {
            freq2list.erase(freq);
            if (min_freq == freq)
                ++min_freq;
        }
    }

    int min_freq = 0, capacity;
    unordered_map<int, list<Node>::iterator> key2node;
    unordered_map<int, list<Node>> freq2list;
};

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache* obj = new LFUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 0460. LFU Cache/Solution.py
from collections import defaultdict


class DLinkedNode:
    __slots__ = "key", "value", "freq", "prev", "next"

    def __init__(self, key: int = 0, value: int = 0) -> None:
        self.key = key
        self.value = value
        self.freq = 1


class LFUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.key2node = {}

        def newDLinkedList() -> DLinkedNode:
            head = DLinkedNode()
            head.prev = head
            head.next = head
            return head

        self.freq2list = defaultdict(newDLinkedList)
        self.min_freq = 1

    def get(self, key: int) -> int:
        if key not in self.key2node:
            return -1
        return self.get_node(key).value

    def put(self, key: int, value: int) -> None:
        if key in self.key2node:
            node = self.get_node(key)
            node.value = value
        else:
            if len(self.key2node) == self.capacity:
                back_node = self.popBack(self.freq2list[self.min_freq])
                self.key2node.pop(back_node.key)
                if self.freq2list[self.min_freq].next == self.freq2list[self.min_freq]:
                    self.freq2list.pop(self.min_freq)
            node = DLinkedNode(key, value)
            self.key2node[key] = node
            self.pushFront(self.freq2list[1], node)
            self.min_freq = 1

    def get_node(self, key: int) -> DLinkedNode:
        node = self.key2node[key]
        node.prev.next = node.next
        node.next.prev = node.prev

        head = self.freq2list[node.freq]
        if head.next == head:
            self.freq2list.pop(node.freq)
            if self.min_freq == node.freq:
                self.min_freq += 1
        node.freq += 1
        self.pushFront(self.freq2list[node.freq], node)
        return node

    def pushFront(self, head: DLinkedNode, node: DLinkedNode) -> None:
        node.next = head.next
        node.prev = head
        head.next = node
        node.next.prev = node

    def popBack(self, head: DLinkedNode) -> DLinkedNode:
        node = head.prev
        node.prev.next = head
        head.prev = node.prev
        return node


# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
